### PR TITLE
BindingContext Unit Tests

### PR DESCRIPTION
**Pull request by alxrsngrtn**
- Broke out DeferredStore into it's own file, added tests for that file
- Unable to test private field that sets the IBinder death actions; requires IPC test setup.
- Created Fakes for tests (cannot use mocks internally due to suspend function).

### CODE REVIEW

File: java/arcs/android/storage/service/DeferredStore.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * This class wraps an [ActiveStore] constructor. The first time an instance of this class is
+ * invoked, the store instance is created.
+ *
+ * This allows us to create a [BindingContext] without blocking the thread that the bind call
+ * occurs on.
+ */
+class DeferredStore<Data : CrdtData, Op : CrdtOperation, T>(
```
Comment (jasonwyatt) [inline java/arcs/android/storage/service/DeferredStore.kt:29]: Refactoring like this as part of writing unit tests should probably be included in the PR description.

**Review by jasonwyatt — APPROVED** [review]

File: java/arcs/android/storage/service/DeferredStore.kt (hunk 1 above)
Comment (alxrsngrtn) [reply java/arcs/android/storage/service/DeferredStore.kt:29]: Noted, thanks!

File: javatests/arcs/android/storage/service/DeferredStoreTest.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ */
+@RunWith(AndroidJUnit4::class)
+@OptIn(ExperimentalCoroutinesApi::class)
+class DeferredStoreTest {
+
+  @Test
+  fun deferredStore_construction_doesNotCreateActiveStore() = runBlockingTest {
+    // We can confirm an active store is not created if these factories / providers are not invoked.
+    val driverFactoryMock = mock<DriverFactory>()
+    val writeBackMock = mock<WriteBack>()
```
Comment (galganif) [inline javatests/arcs/android/storage/service/DeferredStoreTest.kt:47]: nit: normally we'd call these mockDriverFactory and mockWriteBack

**Review by galganif — APPROVED** [review]

File: javatests/arcs/android/storage/service/DeferredStoreTest.kt (hunk 2 above)
Comment (alxrsngrtn) [reply javatests/arcs/android/storage/service/DeferredStoreTest.kt:47]: thanks, noted.

**Review by galganif — APPROVED** [review]

File: java/arcs/core/storage/testutil/FakeActiveStore.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+package arcs.core.storage.testutil
+
+import arcs.core.crdt.CrdtData
+import arcs.core.crdt.CrdtOperation
+import arcs.core.storage.ActiveStore
+import arcs.core.storage.ProxyCallback
+import arcs.core.storage.ProxyMessage
+import arcs.core.storage.StoreOptions
+
+/** [FakeActiveStore] is used to test exception handling for callback registration. */
```
Comment (csilvestrini) [inline java/arcs/core/storage/testutil/FakeActiveStore.kt:21]: This is a little bit strange. Testing what happens when an exception is thrown is totally fine of course, but having that happen as the default behaviour for a fake is quite unexpected (particularly when the class name `FakeActiveStore` gives no indication that it's specifically for testing exceptions).

There's a couple of other ideas you could consider:

1. Add a nullable `exception` field to this class, and have the `on` method either throw the exception it sees there, or return a default int
2. Make this an open class, make the default behaviour for the `on` method to just return some value (e.g. `0`), and then subclass this fake class in the unit test itself to do the exceptional behaviour.

My preference would be for option 2. 

... And in fact we already have a `NoopActiveStore` class that allows you to do that. You can see an example of overriding this very method in `StorageChannelImplTest.kt`

File: java/arcs/core/storage/testutil/FakeDriverFactory.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.storage.testutil
+
+import arcs.core.storage.Driver
+import arcs.core.storage.DriverFactory
+import arcs.core.storage.StorageKey
+import kotlin.reflect.KClass
+
+/** [FakeDriverFactory] is used to control what [Driver] is returned in tests*/
```
Comment (csilvestrini) [inline java/arcs/core/storage/testutil/FakeDriverFactory.kt:18]: Fix punctuation at the end of the line: `tests. */`

File: javatests/arcs/android/storage/service/BindingContextTest.kt [hunk 5]
```diff
@@ -80,9 +85,10 @@ class BindingContextTest {
   }
 
   private fun buildContext(
+    storeProvider: suspend () -> ActiveStore<*, *, *> = { store },
```
Comment (csilvestrini) [inline javatests/arcs/android/storage/service/BindingContextTest.kt:88]: Why not make this `store: ActiveStore<*, *, *> = store`, and then below in line 85 construct the `storeProvider`. That way you make it ever so slightly easier for callers to call this method.

File: javatests/arcs/android/storage/service/BindingContextTest.kt [hunk 6]
```diff
@@ -121,6 +127,24 @@ class BindingContextTest {
...
+    val bindingContext = buildContext(storeProvider = { fakeStore })
+    val callback = DeferredProxyCallback()
+
+    val e = assertFailsWith(CrdtException::class) {
+      suspendForRegistrationCallback {
+        bindingContext.registerCallback(callback, it)
+      }
+    }
+    assertThat(e).hasCauseThat().hasCauseThat()
+      .hasMessageThat().contains("Intentionally thrown!")
```
Comment (csilvestrini) [inline javatests/arcs/android/storage/service/BindingContextTest.kt:145]: This relates to my comment in `FakeActiveStore.kt`: it's not obvious reading this test case where the `"Intentionally thrown!"` error message is coming from. Intentionally thrown from where? You have to know to go digging into the implementation of `FakeActiveStore` to find it. Having to go digging anywhere is not ideal when trying to understand a test case, but particularly not so when the place you need to look isn't the test file nor the class under test. 

This is why I'd suggest to override the `on` method directly in the test case. Then you'll be able to very clearly see where it came from:

```
val fakeStore = object : NoopActiveStore {
  override fun on() = throw BlahBlah()
}

...

assertThat(e).hasCauseThat().isInstanceOf(BlahBlah)
```

File: javatests/arcs/android/storage/service/BindingContextTest.kt [hunk 7]
```diff
@@ -177,3 +201,5 @@ class BindingContextTest {
     assertThat(receivedMessage).isEqualTo(message)
   }
 }
+
+typealias ProxyCallback = suspend (ProxyMessage<*, *, *>) -> Unit
```
Comment (csilvestrini) [inline javatests/arcs/android/storage/service/BindingContextTest.kt:205]: Delete this line? I can't see where it's being used.

File: javatests/arcs/android/storage/service/DeferredStoreTest.kt [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+ */
+@RunWith(AndroidJUnit4::class)
+@OptIn(ExperimentalCoroutinesApi::class)
+class DeferredStoreTest {
+
+  @Test
+  fun deferredStore_construction_doesNotCreateActiveStore() = runBlockingTest {
+    // We can confirm an active store is not created if these factories / providers are not invoked.
+    val mockDriverFactory = mock<DriverFactory>()
+    val mockWriteBack = mock<WriteBack>()
```
Comment (csilvestrini) [inline javatests/arcs/android/storage/service/DeferredStoreTest.kt:46]: The mocks that are shared between tests can be moved to be member variables, then you don't have to redefine them in every test.

File: java/arcs/core/storage/testutil/FakeActiveStore.kt (hunk 3 above)
Comment (alxrsngrtn) [reply java/arcs/core/storage/testutil/FakeActiveStore.kt:21]: Awesome, I'm happy to use something that already exists. Thanks for the suggestions!

File: javatests/arcs/android/storage/service/DeferredStoreTest.kt (hunk 8 above)
Comment (csilvestrini) [reply javatests/arcs/android/storage/service/DeferredStoreTest.kt]: Not done? I meant move the definitions of `mockDriverFactory` and `writeBack` up to be member variables in this test class.

**Review by csilvestrini — APPROVED** [review]

**Review by galganif — APPROVED** [review]

**Pull request closed without merging**